File: 2015/day06/sol.go
package main

import (
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

func main() {
	f, err := os.Open("input")
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	buf, err := io.ReadAll(f)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	lines := strings.Split(string(buf), "\n")
	light1 := [1000][1000]bool{}
	light2 := [1000][1000]int{}
	for _, line := range lines {
		idx := strings.IndexAny(line, "0123456789")
		cmd := line[:idx-1]
		f := strings.Split(line[idx:], " ")
		p1 := strings.Split(f[0], ",")
		p2 := strings.Split(f[2], ",")
		p1x, p1y := atoi(p1[0]), atoi(p1[1])
		p2x, p2y := atoi(p2[0]), atoi(p2[1])
		for y := p1y; y <= p2y; y++ {
			for x := p1x; x <= p2x; x++ {
				switch cmd {
				case "turn on":
					light1[y][x] = true
					light2[y][x]++
				case "toggle":
					light1[y][x] = !light1[y][x]
					light2[y][x] += 2
				case "turn off":
					light1[y][x] = false
					light2[y][x]--
					if light2[y][x] < 0 {
						light2[y][x] = 0
					}
				default:
					panic("unhandled cmd")
				}
			}
		}
	}

	part1, part2 := 0, 0
	for y := range light1 {
		for x := range light1[y] {
			if light1[y][x] {
				part1++
			}
			part2 += light2[y][x]
		}
	}
	fmt.Println("part1:", part1)
	fmt.Println("part2:", part2)
}

func atoi(s string) int {
	n, err := strconv.Atoi(s)
	if err != nil {
		panic(err)
	}
	return n
}

File: 2020/03/sol.py

def main():
	with open("input") as f:
		map = [line.strip() for line in f]
	print("part1:", count(map, 3, 1))
	part2 = 1
	for dx, dy in [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]:
		part2 *= count(map, dx, dy)
	print("part2:", part2)

def count(map, dx, dy):
	out = 0
	x, y = 0, 0
	while y < len(map):
		out += at(map, x, y) == "#"
		x += dx
		y += dy
	return out

def at(map, x, y):
	n = len(map[0])
	return map[y][x%n]

if __name__ == "__main__":
	main()
